### PR TITLE
Refactor Parquet writer options and builders

### PR DESCRIPTION
**Pull request by etseidl**
## Description
Adding options to the Parquet writer is made somewhat tedious by the duplication of code between the two current sets of options/builder classes; one each for the chunked and non-chunked Parquet writers. This PR pulls common options into a parent options class, and common setters into a parent builder class. The builder parent uses CRTP to allow chaining of options. 

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by etseidl** [comment]
cc @vuule @mhaseeb123

**Comment by vyasr** [comment]
Seems closely related to https://github.com/rapidsai/cudf/issues/15825, just for a different format, right? Just wondering if it's worth rolling that change in too; no worries if not.

**Comment by etseidl** [comment]
> Seems closely related to #15825, just for a different format, right? Just wondering if it's worth rolling that change in too; no worries if not.

Yeah, I noticed that issue after submitting this. I can take a look tomorrow.

**Comment by ttnghia** [comment]
Wow, it seems like there are rooms for Parquet reader and ORC reader/writer too :+1:

**Comment by mhaseeb123** [comment]
Thanks for the effort, as always @etseidl. I quickly went over the long-overdue changes and was wondering if we need to update corresponding python bindings in `python.pxd` as well?

**Comment by etseidl** [comment]
> I was wondering if we need to update corresponding python bindings in `python.pxd` and perhaps `python.pyx` as well?

@mhaseeb123 I was worried about that when I started (due to an offline discussion with @vuule), but since the API didn't change, the cython bindings didn't have to change at all 🤯. It all just worked ("worked" == "all tests passed"...just ignore all those CI failures 🤣).

**Comment by etseidl** [comment]
> Wow, it seems like there are rooms for Parquet reader and ORC reader/writer too 👍

Oh dear 😂

**Comment by vyasr** [comment]
> > I was wondering if we need to update corresponding python bindings in `python.pxd` and perhaps `python.pyx` as well?
> 
> @mhaseeb123 I was worried about that when I started (due to an offline discussion with @vuule), but since the API didn't change, the cython bindings didn't have to change at all 🤯. It all just worked ("worked" == "all tests passed"...just ignore all those CI failures 🤣).

While this is true, in the future if you make changes to add to the API of the base class you'll have a similar issue to what this PR is trying to address: you'll have to add those methods to both the chunked and non-chunked versions in Cython because they don't have the concept of the inheritance structure. If it's not too much trouble I would recommend making the changes in this PR to save yourself the trouble. [Here's an example of how that inheritance looks in Cython](https://github.com/rapidsai/cudf/blob/branch-24.06/python/cudf/cudf/_lib/pylibcudf/libcudf/aggregation.pxd#L58).

To give a bit of context, the way that exposing C++ classes to Cython works is that pxd files are basically a promise to Cython that if it generates C++ code calling a certain function, that code will exist. So if you add something to a pxd file but never call that function in any pyx files, the generated C++ code will remain unchanged.

**Comment by etseidl** [comment]
> While this is true, in the future if you make changes to add to the API of the base class you'll have a similar issue to what this PR is trying to address: you'll have to add those methods to both the chunked and non-chunked versions in Cython because they don't have the concept of the inheritance structure. If it's not too much trouble I would recommend making the changes in this PR to save yourself the trouble.

Drat, I thought I'd get away with no python changes :sob:. Thanks for pointing this out.

**Comment by vuule** [comment]
> > Wow, it seems like there are rooms for Parquet reader and ORC reader/writer too 👍
> 
> Oh dear 😂

Yeah! I hope I can count on you to refactor one of these @ttnghia :)

File: cpp/include/cudf/io/parquet.hpp
Comment (mhaseeb123) [inline cpp/include/cudf/io/parquet.hpp:32]: Is this header needed or automatically added by `clangd`?

Comment (etseidl) [reply cpp/include/cudf/io/parquet.hpp:32]: clangd...part of its insidious IWYU campaign

File: cpp/include/cudf/io/parquet.hpp
Comment (vuule) [inline cpp/include/cudf/io/parquet.hpp:925]: should this be a const&?

Comment (etseidl) [reply cpp/include/cudf/io/parquet.hpp:925]: IIRC this was a clangd suggestion along with passing in an initializer list.

**Comment by etseidl** [comment]
@vyasr I've made the python changes, please check them out. I hope I did the CRTP in cython correctly. Also, I just guessed at the formatting since `pre-commit` would only flag problems but not fix them. It seems happy now, but that might be a low bar.

**Comment by vyasr** [comment]
Sorry for the bit of churn there, had to convince GH that everything was up to date as I thought it was after all the merges of upstream branches.

**Review by vyasr — APPROVED** [review]
I've never seen CRTP in Cython before. I'm glad it seemingly works fine! Approving, pending seeing builds and tests all pass; I'm trusting them to verify that Cython won't crash here unexpectedly.

**Comment by etseidl** [comment]
> I've never seen CRTP in Cython before. I'm glad it seemingly works fine! Approving, pending seeing builds and tests all pass; I'm trusting them to verify that Cython won't crash here unexpectedly.

I swear I built this before...but just did a clean build that failed. Not entirely sure what the problem is, but redefining the child methods to return `parquet_writer_options_builder_base.BuilderT&` compiles and tests pass, but it makes me nervous. The one thing that is different from C++ to python is I omit the second template param to `parquet_writer_options_builder_base` since it's only needed for the return type of `build()`, which in cython I keep in the child classes. I'll test and see if that makes a difference. I'm at a loss as to why the code as is fails with `'parquet_writer_options_builder' is not a type identifier`.

**Comment by vyasr** [comment]
Yikes I'm sorry @etseidl I meant to respond earlier and then completely forgot about it. I appreciate how hard you've tried to get the Cython working! I am (unfortunately) not that surprised that it's proving tricky. If the example from the Cython test suite is working for you and seems that similar, my next guess would be some sort of circular dependency issue in our Cython due to how many things we're importing? It's probably not the best use of your time to try and track this down though, so I'm fine with you pushing the code back up to the base class in order to get this merged. Once this PR is merged I can test out the Cython in a follow-up and see what I come up with. Thank you for getting it this far!

**Review by vuule — APPROVED** [review]

**Review by mhaseeb123 — APPROVED** [review]

**Review by hyperbolic2346 — APPROVED** [review]

**Pull request merged**
